scripts/o3de/o3de/engine_properties.py

```diff
@@ -18,25 +18,35 @@
 logger = logging.getLogger('o3de.engine_properties')
 logging.basicConfig(format=utils.LOG_FORMAT)
 
+
 def _edit_gem_names(engine_json: dict,
                     new_gem_names: str or list = None,
                     delete_gem_names: str or list = None,
-                    replace_gem_names: str or list = None):
+                    replace_gem_names: str or list = None,
+                    is_optional: bool = False):
     if new_gem_names:
-        tag_list = new_gem_names.split() if isinstance(new_gem_names, str) else new_gem_names
-        engine_json.setdefault('gem_names', []).extend(tag_list)
+        add_list = new_gem_names.split() if isinstance(new_gem_names, str) else new_gem_names
+        if is_optional:
+            add_list = [dict(name=gem_name, optional=True) for gem_name in add_list]
+        engine_json.setdefault('gem_names', []).extend(add_list)
+
     if delete_gem_names:
         removal_list = delete_gem_names.split() if isinstance(delete_gem_names, str) else delete_gem_names
         if 'gem_names' in engine_json:
-            for tag in removal_list:
-                if tag in engine_json['gem_names']:
-                    engine_json['gem_names'].remove(tag)
+            def in_list(gem: str or dict, remove_list: list) -> bool:
+                if isinstance(gem, dict):
+                    return gem['name'] in remove_list
+                else:
+                    return gem in remove_list
+
+            engine_json['gem_names'] = [gem for gem in engine_json['gem_names'] if not in_list(gem, removal_list)]
+
     if replace_gem_names:
         tag_list = replace_gem_names.split() if isinstance(replace_gem_names, str) else replace_gem_names
         engine_json['gem_names'] = tag_list
 
```

Comment:
We also need to support the `is_optional` field in the replace_gem_names list as well

Reply:
I was thinking that it wouldn't be common to make all the gems optional if replacing the list, which is what this would do.  I'd expect users to replace the list of gems then later do `enable-gem --optional ...` to a smaller set of gems.  What do you think?

Reply:
I don't think it would be common, but it would be surprising to the user that the `--optional` doesn't work when replacing a list of gems though

Reply:
I just looked.  There's no code path that gets there from `enable-gem` sub-command, which is where the `--optional` parameter is added.  The code that it executes is:

```
        ret_val = project_properties.edit_project_props(project_path, new_gem_names=gem_json_data['gem_name'],
                                                                                    is_optional=optional)
```

Enable-gem doesn't have a call to `edit_project_props` that sets `replace_gem_names`, and the `--optional` flag isn't attached to any other user-facing sub-command.

Reply:
I can update both `_edit_gem_names` functions (in project_properties and engine_properties) and add docstring documentation.  That seems like the best place for this information since this function is very internal.

Reply:
But the user can invoke the `edit-project-properties` and `edit-engine-properties` command without going through enable-gem

Reply:
The `--optional` flag is not exposed to the `edit-project-properties` or `edit-engine-properties` sub-commands.  Do you want these to be added?

Reply:
It doesn't need to be now.
If at some point in the future we want to expose it we can do so then.